controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -236,6 +236,44 @@ func (r *StorageClassClaimReconciler) reconcileProviderPhases() (reconcile.Resul
 			r.storageClassClaim.Status.Phase = string(v1alpha1.StorageClassClaimStateReady)
 		}
 
+	} else {
+		r.storageClassClaim.Status.Phase = string(v1alpha1.StorageClassClaimStateDeleting)
+
+		if r.verifyCephResourcesDoNotExist() {
+			r.log.Info("Removing finalizer from StorageClassClaim.", "StorageClassClaim", klog.KRef(r.storageClassClaim.Namespace, r.storageClassClaim.Name))
+			r.storageClassClaim.ObjectMeta.Finalizers = controllers.Remove(r.storageClassClaim.ObjectMeta.Finalizers, storageClassClaimFinalizer)
+			if err := r.update(r.storageClassClaim); err != nil {
+				r.log.Error(err, "Failed to remove finalizer from StorageClassClaim", "StorageClassClaim", klog.KRef(r.storageClassClaim.Namespace, r.storageClassClaim.Name))
+				return reconcile.Result{}, err
+			}
+		} else {
+			for _, cephResource := range r.storageClassClaim.Status.CephResources {
+				switch cephResource.Kind {
+				case "CephClient":
+					cephClient := &rookCephv1.CephClient{}
+					cephClient.Name = cephResource.Name
+					cephClient.Namespace = r.namespace
+					if err := r.delete(cephClient); err != nil {
+						return ctrl.Result{}, fmt.Errorf("unable to delete CephClient : %v", err)
+					}
+
+				case "CephBlockPool":
+					cephBlockPool := &rookCephv1.CephBlockPool{}
+					cephBlockPool.Name = cephResource.Name
+					cephBlockPool.Namespace = r.namespace
+					if err := r.delete(cephBlockPool); err != nil {
+						return ctrl.Result{}, fmt.Errorf("unable to delete CephBlockPool : %v", err)
+					}
+				case "CephFilesystemSubVolumeGroup":
+					cephFilesystemSubVolumeGroup := &rookCephv1.CephFilesystemSubVolumeGroup{}
+					cephFilesystemSubVolumeGroup.Name = cephResource.Name
+					cephFilesystemSubVolumeGroup.Namespace = r.namespace
+					if err := r.delete(cephFilesystemSubVolumeGroup); err != nil {
+						return ctrl.Result{}, fmt.Errorf("unable to delete CephFilesystemSubVolumeGroup : %v", err)
+					}
```

Comment:
Why do we need to delete all of these manually?

First, the block pool and subvolumegroups are shared between multiple claims and should not be deleted until the last claim that is using them gets deleted.

Second, the owner references on the dependent resources should automatically accomplish the garbage collection on these resources when the last owner gets removed. It is a built-in feature of k8s.

Please remove this entire code block!

Reply:
removed